File: java/src/test/java/org/taymyr/lagom/elasticsearch/document/ElasticDocumentIT.java
package org.taymyr.lagom.elasticsearch.document;

import akka.Done;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.taymyr.lagom.elasticsearch.AbstractElasticsearchIT;
import org.taymyr.lagom.elasticsearch.TestDocument;
import org.taymyr.lagom.elasticsearch.TestDocument.IndexedTestDocument;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkRequest;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkResult;
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkUpdate;
import org.taymyr.lagom.elasticsearch.document.dsl.delete.DeleteResult;
import org.taymyr.lagom.elasticsearch.document.dsl.index.IndexResult;
import org.taymyr.lagom.elasticsearch.document.dsl.update.DocUpdateRequest;
import org.taymyr.lagom.elasticsearch.document.dsl.update.UpdateRequest;
import org.taymyr.lagom.elasticsearch.document.dsl.update.UpdateResult;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.taymyr.lagom.elasticsearch.ServiceCall.invoke;

import static java.lang.Thread.sleep;

class ElasticDocumentIT extends AbstractElasticsearchIT {

    @Test
    @DisplayName("Document service descriptor should work correct")
    void shouldWorkCorrect() throws InterruptedException, ExecutionException, TimeoutException {
        indexDocument();
        sleep(1000);
        bulkUpdate();
        sleep(1000);
        getDocument();
        update();
        sleep(1000);
        getSource();
        checkExists();
        delete();
    }

    private void indexDocument() throws InterruptedException, ExecutionException, TimeoutException {
        IndexResult result = eventually(invoke(elasticDocument.indexWithId("test", "sample", "1"),
                new TestDocument("user", "message")));
        assertThat(result.getIndex()).isEqualTo("test");
        assertThat(result.getType()).isEqualTo("sample");
    }

    private void getDocument() throws InterruptedException, ExecutionException, TimeoutException {
        IndexedTestDocument result = eventually(invoke(elasticDocument.get("test", "sample", "1"), IndexedTestDocument.class));
        assertThat(result.getIndex()).isEqualTo("test");
        assertThat(result.getType()).isEqualTo("sample");
        assertThat(result.getId()).isEqualTo("1");
        assertThat(result.getSource().getUser()).isEqualTo("user.bulkUpdate");
        assertThat(result.getSource().getMessage()).isEqualTo("message");
    }

    private void getSource() throws InterruptedException, ExecutionException, TimeoutException {
        TestDocument result = eventually(invoke(elasticDocument.getSource("test", "sample", "1"), TestDocument.class));
        assertThat(result.getUser()).isEqualTo("user.update");
        assertThat(result.getMessage()).isEqualTo("message");
    }

    private void checkExists() throws InterruptedException, ExecutionException, TimeoutException {
        Done result = eventually(elasticDocument.exists("test", "sample", "1").invoke());
        assertThat(result).isEqualTo(Done.getInstance());
        result = eventually(elasticDocument.existsSource("test", "sample", "1").invoke());
        assertThat(result).isEqualTo(Done.getInstance());
    }

    private void bulkUpdate() throws InterruptedException, ExecutionException, TimeoutException {
        BulkRequest request = BulkRequest.of(
                new BulkUpdate("1", new IndexedTestDocument(new TestDocument("user.bulkUpdate", "message")))
        );
        BulkResult result = eventually(elasticDocument.bulk("test", "sample").invoke(request));
        assertThat(result.isErrors()).isFalse();
        assertThat(result.getItems()).hasSize(1);
        assertThat(result.getItems().get(0).getStatus()).isEqualTo(200);
        assertThat(result.getItems().get(0).getResult()).isEqualTo("updated");
        assertThat(result.getItems().get(0).getError()).isNull();
        assertThat(result.getItems().get(0).getIndex()).isEqualTo("test");
        assertThat(result.getItems().get(0).getType()).isEqualTo("sample");
        assertThat(result.getItems().get(0).getId()).isEqualTo("1");
    }

    private void update() throws InterruptedException, ExecutionException, TimeoutException {
        TestDocument doc = new TestDocument("user.update", "message");
        UpdateRequest updateRequest = DocUpdateRequest.builder().doc(doc).build();
        UpdateResult result = eventually(invoke(elasticDocument.update("test", "sample", "1"), updateRequest));
        assertThat(result.getIndex()).isEqualTo("test");
        assertThat(result.getType()).isEqualTo("sample");
        assertThat(result.getId()).isEqualTo("1");
        assertThat(result.getResult()).isEqualTo("updated");
    }

    private void delete() throws InterruptedException, ExecutionException, TimeoutException {
        DeleteResult result = eventually(elasticDocument.delete("test", "sample", "1").invoke());
        assertThat(result.getIndex()).isEqualTo("test");
        assertThat(result.getType()).isEqualTo("sample");
        assertThat(result.getId()).isEqualTo("1");
        assertThat(result.getResult()).isEqualTo("deleted");
    }

}

File: src/test/kotlin/org/taymyr/lagom/elasticsearch/document/ElasticDocumentIT.kt
package org.taymyr.lagom.elasticsearch.document

import akka.Done
import io.kotlintest.extensions.TestListener
import io.kotlintest.matchers.beInstanceOf
import io.kotlintest.matchers.collections.shouldHaveSize
import io.kotlintest.matchers.string.containIgnoringCase
import io.kotlintest.shouldBe
import io.kotlintest.shouldHave
import io.kotlintest.shouldNotBe
import io.kotlintest.specs.WordSpec
import io.kotlintest.whenReady
import org.taymyr.lagom.elasticsearch.IndexedSampleDocument
import org.taymyr.lagom.elasticsearch.LagomClientAndEmbeddedElastic
import org.taymyr.lagom.elasticsearch.SampleDocument
import org.taymyr.lagom.elasticsearch.deser.invoke
import org.taymyr.lagom.elasticsearch.deser.invokeT
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkCreate
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkDelete
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkIndex
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkRequest
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkResult

class ElasticDocumentIT : WordSpec() {

    override fun listeners(): List<TestListener> = listOf(Companion)

    init {
        // TODO(Ilya Korshunov): add tests for update bulk command
        "ElasticDocument" should {
            val document = SampleDocument("user", "message")
            "successfully add a document to index" {
                whenReady(elasticDocument.indexWithId("test", "sample", "1")
                    .invoke(document).toCompletableFuture()) { result ->
                    result.index shouldBe "test"
                    result.type shouldBe "sample"
                }
            }
            "successfully check to exist a document" {
                whenReady(elasticDocument.exists("test", "sample", "1")
                    .invoke().toCompletableFuture()) { result ->
                    result shouldBe Done.getInstance()
                }
            }
            "successfully get a document by id" {
                whenReady(elasticDocument.get("test", "sample", "1")
                    .invokeT<IndexedSampleDocument>().toCompletableFuture()) { result ->
                    result.source shouldBe document
                    result.index shouldBe "test"
                    result.type shouldBe "sample"
                    result.id shouldBe "1"
                    result.version shouldBe 1
                }
            }
            "successfully check to exist a document source" {
                whenReady(elasticDocument.existsSource("test", "sample", "1")
                    .invoke().toCompletableFuture()) { result ->
                    result shouldBe Done.getInstance()
                }
            }
            "successfully get a document source by id" {
                whenReady(elasticDocument.getSource("test", "sample", "1")
                    .invokeT<SampleDocument>().toCompletableFuture()) { result ->
                    result shouldBe document
                }
            }
            val testEntity = IndexedSampleDocument(SampleDocument("tt", "mm"))
            "successfully add a document via bulk" {
                val request = BulkRequest.ofCommands(BulkCreate("12", testEntity))
                whenReady(elasticDocument.bulk("test", "sample").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe false
                    result.items shouldHaveSize 1
                    val item = result.items[0]
                    item.status shouldBe 201
                    item.result shouldBe "created"
                    item.error shouldBe null
                    item.index shouldBe "test"
                    item.type shouldBe "sample"
                    item.id shouldBe "12"
                }
            }
            "error on creating exists document" {
                val request = BulkRequest.ofCommands(BulkCreate("12", testEntity))
                whenReady(elasticDocument.bulk("test", "sample").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe true
                    result.items shouldHaveSize 1
                    val item = result.items[0]
                    item.status shouldBe 409
                    item.error shouldBe beInstanceOf(BulkResult.BulkCommandResult.ResultItemError::class)
                    item.error shouldNotBe null
                    item.error!!.reason shouldHave containIgnoringCase("document already exists")
                    item.error!!.type shouldBe "version_conflict_engine_exception"
                }
            }
            "successful index exists document" {
                val request = BulkRequest.ofCommands(BulkIndex("12", testEntity))
                whenReady(elasticDocument.bulk("test", "sample").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe false
                    result.items shouldHaveSize 1
                    val item = result.items[0]
                    item.status shouldBe 200
                    item.result shouldBe "updated"
                }
            }
            "successful delete document" {
                val request = BulkRequest.ofCommands(BulkDelete("12"))
                whenReady(elasticDocument.bulk("test", "sample").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe false
                    result.items shouldHaveSize 1
                    val item = result.items[0]
                    item.status shouldBe 200
                    item.result shouldBe "deleted"
                }
            }
        }
    }

    companion object : LagomClientAndEmbeddedElastic()
}